File: 0078-subsets/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> res;

    vector<vector<int>> subsets(vector<int>& nums) {
        res.clear();
        vector<int> current;
        backtrack(0, nums, current);
        return res;
    }

private:
    void backtrack(int index, vector<int>& nums, vector<int> current) {
        if (index == (int)nums.size()) {
            res.push_back(current);   // store this subset
            return;
        }

        // 1) Exclude nums[index]
        backtrack(index + 1, nums, current);

        // 2) Include nums[index] (works on a copy)
        current.push_back(nums[index]);
        backtrack(index + 1, nums, current);
    }
};

File: 0078-subsets/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> arr;

    public List<List<Integer>> subsets(int[] nums) {
        arr = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        backtrack(0, nums, current);
        return arr;
    }

    private void backtrack(int index, int[] nums, List<Integer> current) {
        // base case: reached end — add a copy of current subset
        if (index == nums.length) {
            arr.add(new ArrayList<>(current));
            return;
        }
        backtrack(index + 1, nums, current);
        current.add(nums[index]);
        backtrack(index + 1, nums, current);
        current.remove(current.size() - 1);
    }
}

